### ISSUE TITLE
How to watch user changing geofences

### ISSUE THREAD

**Issue by carcam** [transistorsoft/flutter_background_geolocation#24]
Hi there!!

First of all I want to say Thank you for your awesome library!! It works really nice and it helps a lot in all Geo stuff I have tested so far ;)

I'm working in a Flutter project and I plan to use it as my Geofencing library and I have a couple of question in which I could use your help.

My App is creating geofences dinamically based on some website information I fetch using JSON. This fetch is the first thing that happens in my App. Anyway with that info I add the Geofences and according to debugger, everything works fine. Nevertheless using the Android Studio emulator, I cannot trigger onGeofenceChange event just switching gps coordinates (I have also tried loading a gpx file and playing the path). From my understanding, the Geofence watch is always active after enabling the service, but is there something I need to set so that I can execute some code when the user enters in a geofence?.

And last, the idea is that the user gets a notification when he gets into a geofence and I have seen you can set some notification settings in the library configuration but I have not been able to find any information in the docs or in the nice examples you have on github. Is there some code you can show so that I can send notifications using just your plugin? I'm using other custom code to send notifications but I think it will be nice to just use your stuff.

Thank you very much in advance!!

**Comment by christocracy** [comment]
>Nevertheless using the Android Studio emulator, I cannot trigger onGeofenceChange event

- Is your simulator's current location within `geofenceProximityRadius` (default `1000` meters of the geofence(s)?
- Are you aware of the difference between the `#onGeofence` and `#onGeofencesChange` events?

> just switching gps coordinates (I have also tried loading a gpx file and playing the path).

Devices and simulators don't fire geofences from simulated location.

It's always best to test this plugin on a read device.

>. Is there some code you can show so that I can send notifications using just your plugin?

The plugin has nothing to do with showing notifications (the notification you refer is the the "foreground service" notification required by the operating system.  The plugin has no API for posting arbitrary notifications.

Simply use a 3rd-party module specializing in notifications.

**Comment by carcam** [comment]
Hi Chris,
thank you very much for your reply!!
> Is your simulator's current location within geofenceProximityRadius (default 1000 meters of the geofence(s)?

Indeed!! I have made several tests and on app restart the debug window correctly shows the number of geofences I'm in, so that part works fine ;)

> Are you aware of the difference between the #onGeofence and #onGeofencesChange events?

To be honest I am not 100% sure of the differences. I was expecting to learn more with debugging.

> It's always best to test this plugin on a real device.

That's a bummer :(. But I understand it's out of your reach ;)

> The plugin has nothing to do with showing notifications (the notification you refer is the the "foreground service" notification required by the operating system. The plugin has no API for posting arbitrary notifications.

Makes sense to me, but why do you have "channel name" or even "channel icons" properties. I'm new to app development and also new Android user so I saw these properties similar to a third party library I'm using and I thought they might be related. I'll stick to my plan of using this third party for notifications ;)

Thank you very much for your replies!!

**Comment by christocracy** [comment]
#onGeofence is for listening to a triggered geofence event (enter, exit, dwell).

#onGeofencesChange is fired when the list of monitored geofences changed.

>but why do you have "channel name" or even "channel icons" 

Android foreground-services require a persistent notification to be displayed.  All notifications require a notification-channel.

**Comment by carcam** [comment]
Hi Chris,
thank you very much for your reply again!!

> #onGeofence is for listening to a triggered geofence event (enter, exit, dwell).
> #onGeofencesChange is fired when the list of monitored geofences changed.

Crystal clear!! Also I see it's in Documentation. Sorry I was too obtuse and didn't check it :(. I'm really concerned with the geofence thing and did not see any other details.

Also thank you very much for notifications clarification. I have added my code with third party module and it works fine for onLocation event.

Nevertheless I'm still fighting with the geofence debugging. I can understand it does not work correctly with emulator, but I'm a bit puzzled because if I set emulator to use gps coordinates within a geofence I get this in the debugger on APP start:

 TSGeofenceManager found 4/30 within 1693 meters

Which is awesome. But that does not trigger the bg.BackgroundGeolocation.onGeofence(_onGeofence) listener.

Is there a way to trigger geofence listeners from debugging? How do you unit test these (if so)?

**Comment by carcam** [comment]
Hi @christocracy 
I have installed the apk in a live phone and tried to use geofences with no luck. My code is quite easy I think. I just add have this code overriding initstate():

```
@override
  void initState() {
    super.initState();

    bg.BackgroundGeolocation.onLocation(_onLocation);
    bg.BackgroundGeolocation.onGeofence(_onGeofence);
```
And have this for the _onGeofence method:

```
  void _onGeofence(bg.GeofenceEvent event) {
    print('[${bg.Event.GEOFENCE}] - $event');
    _showNotification(title: 'Geofence event', message: '$event');    
  }
```

But no matter what I do I never get the geofence neither in the emulator, neither in the real android device. Can you please point me what I'm doing wrong?

Best!!

**Comment by christocracy** [comment]
I don't see you executing `#startGeofences` (geofence-only tracking) or `#start` (geofence + location tracking).

The plugin doesn't turn on unless you tell it to.

**Comment by christocracy** [comment]
Nor do I see you executing the required `#ready` method

**Comment by carcam** [comment]
@christocracy Thank you very much for your replay. I was using this code:

`
    bg.BackgroundGeolocation.ready(bg.Config(
        desiredAccuracy: bg.Config.DESIRED_ACCURACY_HIGH,
        distanceFilter: 1.0,
        stopOnTerminate: false,
        startOnBoot: true,
        debug: true,
        logLevel: bg.Config.LOG_LEVEL_VERBOSE,
        reset: true,
        autoSync: false,
        geofenceProximityRadius: 150,
        geofenceInitialTriggerEntry: true,
        notificationChannelName: 'myappchannelname'
    )).then((bg.State state) { 
      if (!state.enabled) {
        ////
        // 3.  Start the plugin.
        //
        bg.BackgroundGeolocation.start();
      }
    });
`

And as first it executes the #ready method and then if state is not enabled it executes the #start I thought it was enough (I copied and pasted this from somewhere in your docs).

Nevertheless I have added the startGeofences after the if and I still have no geofences result. I'm in the geofence and it should show a notification as the #geofenceInitialTriggerEntry is set to true, right?

  ```
  bg.BackgroundGeolocation.ready(bg.Config(
        desiredAccuracy: bg.Config.DESIRED_ACCURACY_HIGH,
        distanceFilter: 1.0,
        stopOnTerminate: false,
        startOnBoot: true,
        debug: true,
        logLevel: bg.Config.LOG_LEVEL_VERBOSE,
        reset: true,
        autoSync: false,
        geofenceProximityRadius: 150,
        geofenceInitialTriggerEntry: true,
        notificationChannelName: 'myappchannelname'
    )).then((bg.State state) { 
      if (!state.enabled) {
        ////
        // 3.  Start the plugin.
        //
        bg.BackgroundGeolocation.start();
      }
      bg.BackgroundGeolocation.startGeofences();
    });
```

**Comment by christocracy** [comment]
- Why??  `geofenceProximityRadius: 150,`   The plugin enforces a minimum `1000` so you don't shoot yourself in the foot like you would have been doing if the plugin allowed this.  I don't think you understand what this is.  Read the docs for this.

>Nevertheless I have added the startGeofences after the if and I still have no geofences result

**No.**.  Read the docs for `#startGeofences`

- `#start` = Track geofences + location.
- `#startGeofences` = Track geofences **ONLY**.

You don't use *both*.  You use **one or the other**.

**Comment by christocracy** [comment]
Are you attempting to get geofences to fire from mocked location or simulator?

**Comment by carcam** [comment]
Hi @christocracy,

Sorry for small radius. I used it for live testing on my phone with real GPS coordinates and I set it to a distance I can easily do on foot.

Anyway I have fixed the code and now my ready call looks like this:

 ```
bg.BackgroundGeolocation.ready(bg.Config(
        desiredAccuracy: bg.Config.DESIRED_ACCURACY_HIGH,
        distanceFilter: 1.0,
        stopOnTerminate: false,
        startOnBoot: true,
        debug: true,
        logLevel: bg.Config.LOG_LEVEL_VERBOSE,
        reset: true,
        autoSync: false,
        geofenceProximityRadius: 1000,
        geofenceInitialTriggerEntry: true,
        notificationChannelName: 'MyCoolAppChannel'
    )).then((bg.State state) { 
      bg.BackgroundGeolocation.startGeofences();
    });
```

And I know geofences are correctly added because logs say this:

D/AudioTrack( 4398): stop() called with 20160 frames delivered
I/TSLocationManager( 4398): [c.t.locationmanager.data.a.a a] 
I/TSLocationManager( 4398):   ✅  Colorado
D/TSLocationManager( 4398): [c.t.l.g.TSGeofenceManager$b run] evaluation buffer timer elapsed
D/TSLocationManager( 4398): [c.t.l.g.TSGeofenceManager a] 
D/TSLocationManager( 4398): ╔═════════════════════════════════════════════
D/TSLocationManager( 4398): ║ TSGeofenceManager found 1/1 within 1000 meters
D/TSLocationManager( 4398): ╠═════════════════════════════════════════════
D/TSLocationManager( 4398): ╟─ 🎾  Colorado
D/TSLocationManager( 4398): ╚═════════════════════════════════════════════
I/flutter ( 4398): [addGeofence] success

Also the onLocation event correctly fires and reply with a location within the radius of the fence (actually the log also reflects it) but even though the  geofenceInitialTriggerEntry set to true and the notifyOnExit also set to true, the geofence event is not triggered.

This last test has been done with Android emulator (in computer) and using the andorid emultor GPS coordinates before executing the app. Also I uninstalled it and reinstalled it in the emulator trying to avoid state issues.

Finally when using the app in a real phone (no mocking location, just using different geofences coordinates) I do not get the onlocation neither the ongeofence event triggered (not sure if this is related with me not having a license yet)

I'm really sorry for all the trouble and for asking so much but I had some issues finding all your documentation (now I have found the wiki on github which helps a lot with debugging) and the "advanced app" code does not help me understand fully geofences as to be honest, I have not been even able to make geofences work on my device with your app... Anyway if you know another code using your geofencing system I could try to give it a go and do not bother you so much.

Best!!

**Comment by christocracy** [comment]
Geofences don't fire with simulated location.  Go test on a real device and start walking.

**Comment by christocracy** [comment]
In my demo app, you can "long-press" on map to add a geofence.  You *must* test geofences on a *real device* in the *real world*.

**Comment by carcam** [comment]
Hi @christocracy,
thank you very much for your reply again.

I'm testing in a real android device now. It's an android 7.0 device I got for testing in this project.

I have enabled a "log" screen to see log data in the device and I have noticed  that when using the emulator in the debug screen I can see all the geofences added and a summary like this:


D/TSLocationManager( 7580): ╔═════════════════════════════════════════════
D/TSLocationManager( 7580): ║ TSGeofenceManager found 1/17 within 1000 meters
D/TSLocationManager( 7580): ╠═════════════════════════════════════════════
D/TSLocationManager( 7580): ╟─ 🎾  Colorado
D/TSLocationManager( 7580): ╚═════════════════════════════════════════════

Nevertheless when I check the log in the real device there is no sign of these geofences in any case (see attached).

It's exactly the same code in the emulator and in the device (I compile it using standard *flutter build apk*)

Can you please provide any light into this?

[BackgroundGeolocation log](https://github.com/transistorsoft/flutter_background_geolocation/files/2848033/BackgroundGeolocation.log)

Thank you very much!!

**Comment by christocracy** [comment]
Your log is basically empty and there's no sign that you called `#start` upon the plugin.

**Comment by christocracy** [comment]
Since you've configured `debug: true`, you should *hear* the plugin go *beeeeeeeeep* (long beep) after `#startGeofences`.

When a geofence trigger event occurs, you'll hear sound *beep-beep-beep-beep* (4 beeps).

Make sure your geofence `radius` is `>= 150` meters.

**Comment by carcam** [comment]
Hi @christocracy,
once more thank you very much for your reply!! I do appreciate your help giving I'm a complete newbie on App development.

I was really shocked by the fact the plugin was correctly triggered in emulator but not on apk and I have triggered the app install and run through usb using flutter run and voilá!! Everything worked perfect.

I was using flutter build to create the apk and I have seen in Flutter documentation that that makes flutter build an APK in "release" mode.

So my guess is that as I have not yet bought the license for the library, it was not working in release mode and once I have made it work in debug mode, it worked perfectly.

I thought the license was just to avoid the "no license" message!! 

Anyway thank you very much for your enormous help on this, and for developing such an awesome library which for sure will finally fit my needs and it will be on my final app.

Best!!!
